# Springboard Data Science Career Track Unit 4 Challenge - Tier 3 Complete

## Objectives
Hey! Great job getting through those challenging DataCamp courses. You're learning a lot in a short span of time. 

In this notebook, you're going to apply the skills you've been learning, bridging the gap between the controlled environment of DataCamp and the *slightly* messier work that data scientists do with actual datasets!

Here’s the mystery we’re going to solve: ***which boroughs of London have seen the greatest increase in housing prices, on average, over the last two decades?***


A borough is just a fancy word for district. You may be familiar with the five boroughs of New York… well, there are 32 boroughs within Greater London [(here's some info for the curious)](https://en.wikipedia.org/wiki/London_boroughs). Some of them are more desirable areas to live in, and the data will reflect that with a greater rise in housing prices.

***This is the Tier 3 notebook, which means it's not filled in at all: we'll just give you the skeleton of a project, the brief and the data. It's up to you to play around with it and see what you can find out! Good luck! If you struggle, feel free to look at easier tiers for help; but try to dip in and out of them, as the more independent work you do, the better it is for your learning!***

This challenge will make use of only what you learned in the following DataCamp courses: 
- Prework courses (Introduction to Python for Data Science, Intermediate Python for Data Science)
- Data Types for Data Science
- Python Data Science Toolbox (Part One) 
- pandas Foundations
- Manipulating DataFrames with pandas
- Merging DataFrames with pandas

Of the tools, techniques and concepts in the above DataCamp courses, this challenge should require the application of the following: 
- **pandas**
    - **data ingestion and inspection** (pandas Foundations, Module One) 
    - **exploratory data analysis** (pandas Foundations, Module Two)
    - **tidying and cleaning** (Manipulating DataFrames with pandas, Module Three) 
    - **transforming DataFrames** (Manipulating DataFrames with pandas, Module One)
    - **subsetting DataFrames with lists** (Manipulating DataFrames with pandas, Module One) 
    - **filtering DataFrames** (Manipulating DataFrames with pandas, Module One) 
    - **grouping data** (Manipulating DataFrames with pandas, Module Four) 
    - **melting data** (Manipulating DataFrames with pandas, Module Three) 
    - **advanced indexing** (Manipulating DataFrames with pandas, Module Four) 
- **matplotlib** (Intermediate Python for Data Science, Module One)
- **fundamental data types** (Data Types for Data Science, Module One) 
- **dictionaries** (Intermediate Python for Data Science, Module Two)
- **handling dates and times** (Data Types for Data Science, Module Four)
- **function definition** (Python Data Science Toolbox - Part One, Module One)
- **default arguments, variable length, and scope** (Python Data Science Toolbox - Part One, Module Two) 
- **lambda functions and error handling** (Python Data Science Toolbox - Part One, Module Four) 

## The Data Science Pipeline

This is Tier Three, so we'll get you started. But after that, it's all in your hands! When you feel done with your investigations, look back over what you've accomplished, and prepare a quick presentation of your findings for the next mentor meeting. 

Data Science is magical. In this case study, you'll get to apply some complex machine learning algorithms. But as  [David Spiegelhalter](https://www.youtube.com/watch?v=oUs1uvsz0Ok) reminds us, there is no substitute for simply **taking a really, really good look at the data.** Sometimes, this is all we need to answer our question.

Data Science projects generally adhere to the four stages of Data Science Pipeline:
1. Sourcing and loading 
2. Cleaning, transforming, and visualizing 
3. Modeling 
4. Evaluating and concluding 


### 1. Sourcing and Loading 

Any Data Science project kicks off by importing  ***pandas***. The documentation of this wonderful library can be found [here](https://pandas.pydata.org/). As you've seen, pandas is conveniently connected to the [Numpy](http://www.numpy.org/) and [Matplotlib](https://matplotlib.org/) libraries. 

***Hint:*** This part of the data science pipeline will test those skills you acquired in the pandas Foundations course, Module One. 

#### 1.1. Importing Libraries

In [1]:
# Let's import the pandas, numpy libraries as pd, and np respectively. 
import pandas as pd
import numpy as np

# Load the pyplot collection of functions from matplotlib, as plt 
from matplotlib import pyplot as plt

#### 1.2.  Loading the data
Your data comes from the [London Datastore](https://data.london.gov.uk/): a free, open-source data-sharing portal for London-oriented datasets. 

In [2]:
# First, make a variable called url_LondonHousePrices, and assign it the following link, enclosed in quotation-marks as a string:
# https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls

url_LondonHousePrices = "https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls"

# The dataset we're interested in contains the Average prices of the houses, and is actually on a particular sheet of the Excel file. 
# As a result, we need to specify the sheet name in the read_excel() method.
# Put this data into a variable called properties.  
properties = pd.read_excel(url_LondonHousePrices, sheet_name='Average price', index_col= None)

### 2. Cleaning, transforming, and visualizing
This second stage is arguably the most important part of any Data Science project. The first thing to do is take a proper look at the data. Cleaning forms the majority of this stage, and can be done both before or after Transformation.

The end goal of data cleaning is to have tidy data. When data is tidy: 

1. Each variable has a column.
2. Each observation forms a row.

Keep the end goal in mind as you move through this process, every step will take you closer. 



***Hint:*** This part of the data science pipeline should test those skills you acquired in: 
- Intermediate Python for data science, all modules.
- pandas Foundations, all modules. 
- Manipulating DataFrames with pandas, all modules.
- Data Types for Data Science, Module Four.
- Python Data Science Toolbox - Part One, all modules

**2.1. Exploring your data** 

Think about your pandas functions for checking out a dataframe. 

In [3]:
properties.head()

Unnamed: 0 City of London Barking & Dagenham       Barnet       Bexley  \
0        NaT      E09000001          E09000002    E09000003    E09000004   
1 1995-01-01    91448.98487         50460.2266  93284.51832  64958.09036   
2 1995-02-01    82202.77314        51085.77983  93190.16963  64787.92069   
3 1995-03-01    79120.70256        51268.96956  92247.52435  64367.49344   
4 1995-04-01    77101.20804        53133.50526  90762.87492  64277.66881   

         Brent      Bromley       Camden      Croydon       Ealing  ...  \
0    E09000005    E09000006    E09000007    E09000008    E09000009  ...   
1  71306.56698  81671.47692  120932.8881  69158.16225  79885.89069  ...   
2  72022.26197  81657.55944  119508.8622  68951.09542  80897.06551  ...   
3  72015.76274  81449.31143  120282.2131  68712.44341  81379.86288  ...   
4  72965.63094  81124.41227   120097.899  68610.04641  82188.90498  ...   

    NORTH WEST YORKS & THE HUMBER EAST MIDLANDS WEST MIDLANDS EAST OF ENGLAND  \
0    E12000002          E12000003     E12000004     E12000005       E12000006   
1  43958.48001        44803.42878   45544.52227   48527.52339      56701.5961   
2  43925.42289        44528.80721   46051.57066   49341.29029     56593.59475   
3   44434.8681        45200.46775   45383.82395   49442.17973     56171.18278   
4   44267.7796        45614.34341   46124.23045   49455.93299     56567.89582   

        LONDON   SOUTH EAST   SOUTH WEST Unnamed: 47      England  
0    E12000007    E12000008    E12000009         NaN    E92000001  
1  74435.76052  64018.87894   54705.1579         NaN  53202.77128  
2  72777.93709  63715.02399  54356.14843         NaN   53096.1549  
3  73896.84204  64113.60858  53583.07667         NaN   53201.2843  
4  74455.28754  64623.22395  54786.01938         NaN   53590.8548  

[5 rows x 49 columns]

**2.2. Cleaning the data**

You might find you need to transpose your dataframe, check out what its row indexes are, and reset the index. You  also might find you need to assign the values of the first row to your column headings  . (Hint: recall the .columns feature of DataFrames, as well as the iloc[] method).

Don't be afraid to use StackOverflow for help  with this.

In [4]:
properties = properties.transpose()

In [5]:
properties = properties[:33]

In [6]:
properties = properties.reset_index()

In [7]:
properties.columns = properties.iloc[0]

In [8]:
properties = properties.drop(0)

In [9]:
properties.head()

0          Unnamed: 0        NaT 1995-01-01 00:00:00 1995-02-01 00:00:00  \
1      City of London  E09000001         91448.98487         82202.77314   
2  Barking & Dagenham  E09000002          50460.2266         51085.77983   
3              Barnet  E09000003         93284.51832         93190.16963   
4              Bexley  E09000004         64958.09036         64787.92069   
5               Brent  E09000005         71306.56698         72022.26197   

0 1995-03-01 00:00:00 1995-04-01 00:00:00 1995-05-01 00:00:00  \
1         79120.70256         77101.20804         84409.14932   
2         51268.96956         53133.50526         53042.24852   
3         92247.52435         90762.87492         90258.00033   
4         64367.49344         64277.66881         63997.13588   
5         72015.76274         72965.63094         73704.04743   

0 1995-06-01 00:00:00 1995-07-01 00:00:00 1995-08-01 00:00:00  ...  \
1         94900.51244         110128.0423         112329.4376  ...   
2         53700.34831         52113.12157         52232.19868  ...   
3         90107.23471         91441.24768         92361.31512  ...   
4         64252.32335         63722.70055         64432.60005  ...   
5         74310.48167         74127.03788          73547.0411  ...   

0 2022-05-01 00:00:00 2022-06-01 00:00:00 2022-07-01 00:00:00  \
1          795174.941           817653.89         823465.9092   
2         334860.1394         344686.2927         348768.9455   
3         584042.5146         589629.3011         590956.5544   
4         385259.9068         388894.2716         394492.2897   
5         529771.7571         532821.5739         549660.5579   

0 2022-08-01 00:00:00 2022-09-01 00:00:00 2022-10-01 00:00:00  \
1         867266.9155         863879.8209          902309.282   
2         350456.7649         349749.4368         353520.9551   
3         601802.5584         607196.4449         605114.2948   
4         400778.7012          402890.116         406868.9186   
5         560279.3404         580872.8243         577222.7821   

0 2022-11-01 00:00:00 2022-12-01 00:00:00 2023-01-01 00:00:00  \
1         958869.7855         929060.0982         978120.5133   
2         356902.8691         356645.1966         352043.1242   
3         596539.3541         596637.6706         598835.4533   
4         405329.0206         406858.5581         399887.1166   
5         572466.8925         564118.4765         560637.6337   

0 2023-02-01 00:00:00  
1         960867.8271  
2         344518.0012  
3         602510.1509  
4         399578.3604  
5         548074.5348  

[5 rows x 340 columns]

**2.3. Cleaning the data (part 2)**

You might we have to **rename** a couple columns. How do you do this? The clue's pretty bold...

In [10]:
properties = properties.rename(columns={'Unnamed: 0': 'Boroughs', pd.NaT: 'ID'})

In [11]:
properties.head()

0            Boroughs         ID 1995-01-01 00:00:00 1995-02-01 00:00:00  \
1      City of London  E09000001         91448.98487         82202.77314   
2  Barking & Dagenham  E09000002          50460.2266         51085.77983   
3              Barnet  E09000003         93284.51832         93190.16963   
4              Bexley  E09000004         64958.09036         64787.92069   
5               Brent  E09000005         71306.56698         72022.26197   

0 1995-03-01 00:00:00 1995-04-01 00:00:00 1995-05-01 00:00:00  \
1         79120.70256         77101.20804         84409.14932   
2         51268.96956         53133.50526         53042.24852   
3         92247.52435         90762.87492         90258.00033   
4         64367.49344         64277.66881         63997.13588   
5         72015.76274         72965.63094         73704.04743   

0 1995-06-01 00:00:00 1995-07-01 00:00:00 1995-08-01 00:00:00  ...  \
1         94900.51244         110128.0423         112329.4376  ...   
2         53700.34831         52113.12157         52232.19868  ...   
3         90107.23471         91441.24768         92361.31512  ...   
4         64252.32335         63722.70055         64432.60005  ...   
5         74310.48167         74127.03788          73547.0411  ...   

0 2022-05-01 00:00:00 2022-06-01 00:00:00 2022-07-01 00:00:00  \
1          795174.941           817653.89         823465.9092   
2         334860.1394         344686.2927         348768.9455   
3         584042.5146         589629.3011         590956.5544   
4         385259.9068         388894.2716         394492.2897   
5         529771.7571         532821.5739         549660.5579   

0 2022-08-01 00:00:00 2022-09-01 00:00:00 2022-10-01 00:00:00  \
1         867266.9155         863879.8209          902309.282   
2         350456.7649         349749.4368         353520.9551   
3         601802.5584         607196.4449         605114.2948   
4         400778.7012          402890.116         406868.9186   
5         560279.3404         580872.8243         577222.7821   

0 2022-11-01 00:00:00 2022-12-01 00:00:00 2023-01-01 00:00:00  \
1         958869.7855         929060.0982         978120.5133   
2         356902.8691         356645.1966         352043.1242   
3         596539.3541         596637.6706         598835.4533   
4         405329.0206         406858.5581         399887.1166   
5         572466.8925         564118.4765         560637.6337   

0 2023-02-01 00:00:00  
1         960867.8271  
2         344518.0012  
3         602510.1509  
4         399578.3604  
5         548074.5348  

[5 rows x 340 columns]

**2.4.Transforming the data**

Remember what Wes McKinney said about tidy data? 

You might need to **melt** your DataFrame here. 

In [12]:
properties = pd.melt(properties, id_vars=['Boroughs', 'ID'], var_name='Date', value_name='Average_Price')

Remember to make sure your column data types are all correct. Average prices, for example, should be floating point numbers... 

In [13]:
properties.head()

Boroughs         ID       Date Average_Price
0      City of London  E09000001 1995-01-01   91448.98487
1  Barking & Dagenham  E09000002 1995-01-01    50460.2266
2              Barnet  E09000003 1995-01-01   93284.51832
3              Bexley  E09000004 1995-01-01   64958.09036
4               Brent  E09000005 1995-01-01   71306.56698

**2.5. Cleaning the data (part 3)**

Do we have an equal number of observations in the ID, Average Price, Month, and London Borough columns? Remember that there are only 32 London Boroughs. How many entries do you have in that column? 

Check out the contents of the London Borough column, and if you find null values, get rid of them however you see fit. 

In [14]:
print(properties['Boroughs'].isnull().sum())

0


In [15]:
properties

Boroughs         ID       Date Average_Price
0          City of London  E09000001 1995-01-01   91448.98487
1      Barking & Dagenham  E09000002 1995-01-01    50460.2266
2                  Barnet  E09000003 1995-01-01   93284.51832
3                  Bexley  E09000004 1995-01-01   64958.09036
4                   Brent  E09000005 1995-01-01   71306.56698
...                   ...        ...        ...           ...
10811           Southwark  E09000028 2023-02-01   539637.6331
10812              Sutton  E09000029 2023-02-01   447798.8127
10813       Tower Hamlets  E09000030 2023-02-01    502932.341
10814      Waltham Forest  E09000031 2023-02-01   509781.8616
10815          Wandsworth  E09000032 2023-02-01   639194.1777

[10816 rows x 4 columns]

**2.6. Visualizing the data**

To visualize the data, why not subset on a particular London Borough? Maybe do a line plot of Month against Average Price?

In [16]:
properties['Date'] = properties['Date'].astype('datetime64[ns]')

In [17]:
properties.loc[:, 'Year'] = properties['Date'].dt.year
properties.loc[:, 'Month'] = properties['Date'].dt.month

In [18]:
properties = properties[['Boroughs','Year','Month','Average_Price']]

To limit the number of data points you have, you might want to extract the year from every month value your *Month* column. 

To this end, you *could* apply a ***lambda function***. Your logic could work as follows:
1. look through the `Month` column
2. extract the year from each individual value in that column 
3. store that corresponding year as separate column. 

Whether you go ahead with this is up to you. Just so long as you answer our initial brief: which boroughs of London have seen the greatest house price increase, on average, over the past two decades? 

In [19]:
yearly_avg_price = properties.groupby(['Boroughs', 'Year'])['Average_Price'].mean().reset_index()

In [20]:
price_increase = yearly_avg_price.groupby('Boroughs')['Average_Price'].agg(lambda x: x.iloc[-6] - x.iloc[3]).reset_index()

In [21]:
price_increase = price_increase.rename(columns={'Average_Price': 'Price_Increase'})

In [22]:
price_increase = price_increase.sort_values('Price_Increase', ascending = False).reset_index(drop = True)

In [23]:
price_increase

Boroughs  Price_Increase
0   Kensington & Chelsea    1.095530e+06
1                 Camden    6.717214e+05
2         City of London    6.367770e+05
3   Hammersmith & Fulham    5.501844e+05
4              Islington    5.156751e+05
5   Richmond upon Thames    4.897094e+05
6             Wandsworth    4.712423e+05
7                Hackney    4.567709e+05
8               Haringey    4.441408e+05
9              Southwark    4.128302e+05
10                Barnet    4.109071e+05
11               Lambeth    4.083573e+05
12                Merton    4.042440e+05
13                 Brent    3.921528e+05
14  Kingston upon Thames    3.806912e+05
15                Ealing    3.687074e+05
16        Waltham Forest    3.653019e+05
17                Harrow    3.511292e+05
18         Tower Hamlets    3.499954e+05
19              Lewisham    3.360899e+05
20               Bromley    3.351235e+05
21             Redbridge    3.280767e+05
22            Hillingdon    3.125900e+05
23             Greenwich    3.068114e+05
24               Enfield    3.036015e+05
25              Hounslow    2.968851e+05
26                Newham    2.913512e+05
27                Sutton    2.871755e+05
28              Havering    2.846467e+05
29               Croydon    2.821553e+05
30                Bexley    2.619716e+05
31    Barking & Dagenham    2.349109e+05

**3. Modeling**

Consider creating a function that will calculate a ratio of house prices, comparing the price of a house in 2018 to the price in 1998.

Consider calling this function create_price_ratio.

You'd want this function to:
1. Take a filter of dfg, specifically where this filter constrains the London_Borough, as an argument. For example, one admissible argument should be: dfg[dfg['London_Borough']=='Camden'].
2. Get the Average Price for that Borough, for the years 1998 and 2018.
4. Calculate the ratio of the Average Price for 1998 divided by the Average Price for 2018.
5. Return that ratio.

Once you've written this function, you ultimately want to use it to iterate through all the unique London_Boroughs and work out the ratio capturing the difference of house prices between 1998 and 2018.

Bear in mind: you don't have to write a function like this if you don't want to. If you can solve the brief otherwise, then great! 

***Hint***: This section should test the skills you acquired in:
- Python Data Science Toolbox - Part One, all modules

In [24]:
yearly_avg_price

Boroughs  Year  Average_Price
0    Barking & Dagenham  1995   51817.969390
1    Barking & Dagenham  1996   51718.192690
2    Barking & Dagenham  1997   55974.262309
3    Barking & Dagenham  1998   60285.821083
4    Barking & Dagenham  1999   65320.934441
..                  ...   ...            ...
923          Wandsworth  2019  588496.361833
924          Wandsworth  2020  611610.794833
925          Wandsworth  2021  611512.793250
926          Wandsworth  2022  636425.687497
927          Wandsworth  2023  640755.083300

[928 rows x 3 columns]

In [25]:
avg_price_1998 = yearly_avg_price.loc[yearly_avg_price['Year'] == 1998, 'Average_Price'].astype(float).reset_index(drop=True)
avg_price_2018 = yearly_avg_price.loc[yearly_avg_price['Year'] == 2018, 'Average_Price'].astype(float).reset_index(drop=True)

In [26]:
ratio = avg_price_2018 / avg_price_1998 * 100 - 100

In [27]:
price_increase['Price_1998'] = avg_price_1998
price_increase['Price_2018'] = avg_price_2018
price_increase['Ratio%'] = ratio

In [28]:
price_increase = price_increase.sort_values('Ratio%', ascending = False).reset_index(drop = True)

In [29]:
price_increase = price_increase[['Boroughs','Price_1998','Price_2018','Price_Increase','Ratio%']]

In [30]:
price_increase

Boroughs     Price_1998    Price_2018  Price_Increase  \
0                Lambeth   87869.535804  5.446405e+05    4.083573e+05   
1                 Bexley   75557.474132  4.408594e+05    2.619716e+05   
2                 Sutton   91405.187779  5.042354e+05    2.871755e+05   
3             Hillingdon   75539.051649  4.116290e+05    3.125900e+05   
4                Enfield   67671.270763  3.590225e+05    3.036015e+05   
5             Wandsworth  148031.894542  7.848089e+05    4.712423e+05   
6                  Brent  107419.850108  5.515606e+05    3.921528e+05   
7               Lewisham  268350.122142  1.363880e+06    3.360899e+05   
8              Redbridge  103179.118943  5.115364e+05    3.280767e+05   
9   Richmond upon Thames  170688.973933  8.424104e+05    4.897094e+05   
10  Kensington & Chelsea   60285.821083  2.951967e+05    1.095530e+06   
11  Hammersmith & Fulham  100692.590417  4.928454e+05    5.501844e+05   
12         Tower Hamlets  134148.973300  6.498241e+05    3.499954e+05   
13                Barnet   81532.933366  3.883443e+05    4.109071e+05   
14    Barking & Dagenham  125406.807233  5.966491e+05    2.349109e+05   
15             Greenwich  108049.855117  5.122939e+05    3.068114e+05   
16               Croydon   96505.166407  4.465006e+05    2.821553e+05   
17              Hounslow   94514.395822  4.225911e+05    2.968851e+05   
18                Camden  122359.468033  5.332665e+05    6.717214e+05   
19                Ealing   85602.087977  3.702488e+05    3.687074e+05   
20              Haringey  111343.154650  4.800505e+05    4.441408e+05   
21               Bromley  116399.769158  4.970910e+05    3.351235e+05   
22             Southwark   93030.224322  3.966317e+05    4.128302e+05   
23        City of London   80632.020822  3.426036e+05    6.367770e+05   
24               Hackney   88143.228830  3.702985e+05    4.567709e+05   
25        Waltham Forest   97676.048952  4.102661e+05    3.653019e+05   
26                Merton  175340.921250  7.255253e+05    4.042440e+05   
27              Havering   92087.045921  3.792626e+05    2.846467e+05   
28             Islington  108286.520467  4.434100e+05    5.156751e+05   
29  Kingston upon Thames  114778.250067  4.659075e+05    3.806912e+05   
30                Newham  162956.066025  6.526654e+05    2.913512e+05   
31                Harrow   99746.069682  3.966312e+05    3.511292e+05   

        Ratio%  
0   519.828556  
1   483.475581  
2   451.648530  
3   444.922104  
4   430.539044  
5   430.162038  
6   413.462496  
7   408.246507  
8   395.775116  
9   393.535341  
10  389.661861  
11  389.455450  
12  384.404801  
13  376.303635  
14  375.770935  
15  374.127331  
16  362.670104  
17  347.118201  
18  335.819592  
19  332.523037  
20  331.145090  
21  327.054952  
22  326.347158  
23  324.897705  
24  320.110028  
25  320.027308  
26  313.779810  
27  311.852261  
28  309.478469  
29  305.919643  
30  300.516190  
31  297.640911

### 4. Conclusion
What can you conclude? Type out your conclusion below. 

Look back at your notebook. Think about how you might summarize what you have done, and prepare a quick presentation on it to your mentor at your next meeting. 

We hope you enjoyed this practical project. It should have consolidated your data hygiene and pandas skills by looking at a real-world problem involving just the kind of dataset you might encounter as a budding data scientist. Congratulations, and looking forward to seeing you at the next step in the course! 

### Based on the analysis of the London housing data from 1998 to 2018, it was found that the average house prices for Greater London Boroughs increased significantly over the two decades. Hillingdon has the highest increase ratio at 519.83%, while Waltham Forest has the lowest increase ratio at 297.64%. Kensington & Chelsea has the highest increase amount at £1,095,530, while Barking & Dagenham has the lowest increase amount at £234,910.